compiler-hacking.md

```diff
@@ -0,0 +1,473 @@
+doesn't need it.
+
+## Compiler Mechanics
+
+The compiler operates in a few phases:
+
+1. Parse the file into a parse tree, rooted as a `ProgramNode`.
+2. Resolve input formats, actions, functions, and `Define`/call references.
+3. Resolve variables.
+4. Check the type and stream purity of everything.
+5. Convert everything to bytecode.
+
+There are a lot of classes in the compiler, but the major ones are:
+
+- `OliveNode`, which declares each top-level element in the file (function,
+  olive, define olive, constant)
+- `OliveClauseNode`, which defines clauses in an olive
+- `ExpressionNode`, which defines all the bits and pieces of expressions
+
+`ExperssionNode` has many similar specialisations. For instance, there is an
+expression node for `For` and then ones for each of the sources, clauses,
+subsamples, and collectors in a `For` expression, containing other expressions.
+While these are structurally different, they are built very similarly to `ExpressionNode`.
+
+### Type Information
+Computing captures is done as part of bytecode production. The type system for
+Shesmu is meant to be very simple and static. There is no inheritance and no
+automatic conversion. It's also entirely bottom up. For instance, in Java, type
+information flows both ways:
+
+     List<Foo> x = new ArrayList<>();
+
+Type information about the type argument `Foo` is flowing in the opposite
+direction as the type information about `ArrayList<Foo>` being assigned to
+`List<Foo>`. 
+
+In Shesmu, everything flows from the leaves of the tree towards the root. For instance:
+
+     path_file(dir + 'x.txt')
+
+If `dir` cannot be resolved or was defined by an expression that has a type
+error, this has no valid type. Even though we know that `path_file` expects a
+`path` type and the `+` operator can combine two paths. Shesmu will simply give
+up. This is meant to keep the compiler simple.
+
+### Bytecode Generation
+When generating bytecode, Shesmu uses `GeneratorAdapter` from the ASM package
+to write bytecode. It wraps this in a `Renderer` class. `Renderer` provides a
+number of utility functions, such as inserting `Imyhat` objects into the
+bytecode. It's real purpose is to track extra information. It knows where the
+current stream value is stored, which will be different through many layers of
+capturing. It also tracks how to access signatures.
+
+### Stream Purity
+Shesmu has two kinds of stream purity: olive-level (type) and `For`-level
+(order). They are conceptually similar, but perform different functions.
+
+For all operations on Java streams, the order or type of the items can be
+changed by some operations.
+
+In the case of olives, `Group`, `Join`, `LeftJoin` change the type of the
+stream value while `Where`, `Monitor`, `Reject`, and `Pick` do not. A `Define`
+olive is written expecting the original type of the input stream. Therefore, no
+type-changing operation can precede a call, but the operations can. Therefore,
```

Comment:
"Therefore, no type-changing operation can precede a call, but the operations can." I'm not quite understanding this...

Reply:
Yeah, that was missing some adjectives. :dizzy_face: